lib/Conversion/TorchToLinalg/Utils.cpp

```diff
@@ -67,6 +69,42 @@ Value torch_to_linalg::getZeroPaddedTensor(
   return getPaddedTensor(op, b, input, paddingInts, paddingInts, c0);
 }
 
+// Helper function that adds dynamic padding to a tensor, ignoring unpaddedDims
+// dimensions at the beginning. The high and low padding are the same, and the
+// padding value is zero.
+Value torch_to_linalg::getDynamicZeroPaddedTensor(
+    Operation *op, OpBuilder &b, Value &input, SmallVectorImpl<Value> &padding,
+    int unpaddedDims) {
+  assert(input.getType().isa<RankedTensorType>() &&
+         "input must be RankedTensorType");
+  Location loc = op->getLoc();
+
+  Value c1 = b.create<arith::ConstantOp>(loc, b.getIndexAttr(1));
+  SmallVector<Value> inputDims = getTensorSizes(b, loc, input);
+
+  Value c0 = b.create<arith::ConstantOp>(loc, b.getI64IntegerAttr(0));
+  SmallVector<Value> paddingIncludingUnchanged{c0, c0};
+  paddingIncludingUnchanged.append(padding);
+  for (auto pad = paddingIncludingUnchanged.begin();
+       pad < paddingIncludingUnchanged.end(); pad++)
+    *pad = castIntToIndex(b, loc, *pad);
+
+  auto inputRank = input.getType().cast<RankedTensorType>().getRank();
+  Type elementType = input.getType().cast<RankedTensorType>().getElementType();
+  Type inputType = RankedTensorType::get(
+      llvm::ArrayRef<int64_t>(SmallVector<int64_t>(inputRank, kUnknownSize)),
+      elementType);
+  input = b.create<tensor::CastOp>(loc, inputType, input);
```

Comment:
The first argument of `createPadScalarOp` is the result type so I don't think you need to cast the input to result type.  You could follow the example here https://github.com/llvm/torch-mlir/blob/9208bf0eb63e3589b54b23324b5202916436a220/lib/Conversion/TorchToLinalg/Utils.cpp#L42-L44 to deduce the result type.

Reply:
inferResultType uses ints, so I don't think that's possible here.